### DISCUSSION TITLE
Cannot understand the AllowedScopes options

### DISCUSSION THREAD

**Discussion by andreuwz** [ThreeMammals/Ocelot#1739] [Q&A] [answered]
Hello.

I have the following access token, which includes scopes (<ApiScope>), issued by Identity Server 4, like this:
`"scope": [
    "Cart.API",
    "Catalogue.API",
    "Identity.API",
    "offline_access"
  ]`
Also, in my ocelot configuration, for certain re-routings I have the following configuration:
`
"AllowedScopes": [
          "Cart.API",
          "Catalogue.API",
          "Identity.API",
          "offline_access"
        ]
`

I want to ask, are the scopes created by IS4 the ones Ocelot looks for?
Because I tested the scopes in the token against 2 examples:
`
"AllowedScopes": [
          "Cart.API",
          "Catalogue.API",
          "Identity.API",
          "offline_access",
           "test"
        ]
`
`
"AllowedScopes": [
          "test"
        ]
`
In the first example where test is included in the list of scopes, using the token from above - I get forbidden. If I use only "test" scope I get Unauthorized. Is this expected behavior?

**Comment by ggnaegi** [comment]
@andreuwz As far as I understand it, the second example is giving you the expected result, since the test scope isn't included in the token's scopes.
https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/Authorization/ScopesAuthorizer.cs
```c#
 if (!matchesScopes.Any())
            {
                return new ErrorResponse<bool>(
                    new ScopeNotAuthorizedError($"no one user scope: '{string.Join(',', userScopes)}' match with some allowed scope: '{string.Join(',', routeAllowedScopes)}'"));
            }
```
as for the first example, the error must be somewhere else, since you have some matching scopes (by the way it's an interesting debate: https://github.com/ThreeMammals/Ocelot/issues/231)

**Reply by raman-m** [reply]
Interestingly, a similar question by an author was raised in PR #1520, which was closed, yet the documentation was updated here:
https://github.com/ThreeMammals/Ocelot/blob/ce0227c059f220761acc2d880e554174bdb3c544/docs/features/authentication.rst?plain=1#L192
This has led to an inconsistency due to the terms "all" and "at least." The underlying issue in the user scenario involves matching scopes by the `IScopesAuthorizer` service, as seen in issue #2177 
- #2177

Consequently, I have assigned high priority to this issue and linked it to PR #1431.
Also, seems we need to extend `AllowedScopes` options proposing a setting for matching strategy.

**Comment by raman-m** [comment, marked as answer]
@andreuwz 

> I want to ask, are the scopes created by IS4 the ones Ocelot looks for?
> Because I tested the scopes in the token against 2 examples:
> `
> "AllowedScopes": [
>           "Cart.API",
>           "Catalogue.API",
>           "Identity.API",
>           "offline_access",
>            "test"
>         ]
> `
> `
> "AllowedScopes": [
>           "test"
>         ]
> `
> In the first example where test is included in the list of scopes, using the token from above - I get forbidden. If I use only "test" scope I get Unauthorized. Is this expected behavior?
> 

Andrey, allow me to clarify the "forbidden issue" in the first case. The issue stems from the `IScopesAuthorizer` implementation, which is utilized during the authentication phase in Ocelot. More details about the root cause can be found in issue #2177.

Ultimately, it appears that the "AllowedScopes" feature may not function correctly due to the character-separated `scope` claim within the token. Therefore, in the first scenario, writing multiple roles/scopes has no impact due to bug #2177 (lack of complete logic). It seems the `scope` claim value is either `Cart.API Catalogue.API Identity.API offline_access test` or comma-separated, with the order being irrelevant.

In the second case, an unauthorized status indicates that the roles derived from the `scope` claim were not correctly parsed into the list. The claim value `role1 role2 test` was parsed as a single role `role1 role2 test`, whereas the `ClaimsAuthorizer` fails to find the `test` role during the second authorization stage in `AuthorizationMiddleware`, even though the first authentication stage was successful.

So, all these issues will be addressed in
- #2177 
- #1431
